### PR TITLE
Force redrawing once the scale factor has changed

### PR DESCRIPTION
**Pull request by NenadBojkovski**

### CODE REVIEW

File: src/openfl/_internal/renderer/opengl/GLShape.hx [hunk 1]
```diff
@@ -59,7 +59,7 @@ class GLShape {
 				renderSession.shaderManager.setShader (shader);
 				
 				shader.data.uImage0.input = graphics.__bitmap;
-				shader.data.uImage0.smoothing = renderSession.allowSmoothing;
+				shader.data.uImage0.smoothing = false;
```
Comment (nadako) [inline src/openfl/_internal/renderer/opengl/GLShape.hx]: This doesn't seem to be related to the PR topic? Also does this matter? Because shapes are always rendered scaled, right?

File: src/openfl/display/Stage.hx [hunk 2]
```diff
@@ -1920,6 +1920,18 @@ class Stage extends DisplayObjectContainer implements IModule {
 			
 		}
 		
+		if (__contentsScaleFactor != window.scale && __renderer != null) {
+			
+			__contentsScaleFactor = window.scale;
+			
+			#if (js && html5)
```
Comment (nadako) [inline src/openfl/display/Stage.hx:1927]: Why the `#if` fencing? Shouldn't we do the same for cairo, etc?

File: src/openfl/_internal/renderer/opengl/GLShape.hx (hunk 1 above)
Comment (Neu2o) [reply src/openfl/_internal/renderer/opengl/GLShape.hx]: This is to avoid blurry text rendering, especially when scaling down. Text is rendered on canvas including smoothing if needed, so the canvas don't need to be rendered smoothed again on GL level. This is actually true for all canvases. It is similar to the cache as bitmap problem we had in the past.

But I agree, it is better, also for upstream pull-requests, if we don't mix this.

File: src/openfl/display/Stage.hx (hunk 2 above)
Comment (Neu2o) [reply src/openfl/display/Stage.hx:1927]: I agree, but I think this is done because we have not tested or focused on Cairo much and have not changed all Cairo renderers to `pixelRatio` yet. @NenadBojkovski, can you confirm this?

For example I found this one here: https://github.com/innogames/openfl/pull/53/files#diff-c1a31d088708f6d62a0bd6ffa0a4fa16

```
#if (js && html5)
    CanvasRenderer.scale = stage.window.scale;
#end
```

it was changed to:

```
#if (js && html5)
    renderSession.pixelRatio = stage.window.scale;
#end
```

That's why the fencing is there right?

File: src/openfl/display/SimpleButton.hx [hunk 3]
```diff
@@ -101,6 +101,15 @@ class SimpleButton extends InteractiveObject {
 	}
 	
 	
+	private override function __forceRenderDirty ():Void {
+		
+		super.__forceRenderDirty();
+		
+		__currentState.__forceRenderDirty();
+		
+	}
```
Comment (Neu2o) [inline src/openfl/display/SimpleButton.hx]: Many of the `SimpleButton` functions handle the `_currentState`, but also the `hitTestState` and check first if the states exist. See `setTransformDirty` as an example: https://github.com/innogames/openfl/blob/d8115491b63763ca63c8727c01f3851f5f898917/src/openfl/display/SimpleButton.hx#L378

Should we do it in a similar way?

File: src/openfl/_internal/renderer/opengl/GLShape.hx (hunk 1 above)
Comment (NenadBojkovski) [reply src/openfl/_internal/renderer/opengl/GLShape.hx]: I was also thinking should I put it separate on in this PR. Decided to put it at this PR because it is required to have sharp rendering when zoomed out. Of course I can split it.

File: src/openfl/display/Stage.hx (hunk 2 above)
Comment (NenadBojkovski) [reply src/openfl/display/Stage.hx:1927]: Yes we currently are using pixelRatio only for html5 and js target and that's way is fenced. Once we(or someone else) implement support for cairo we can remove it

File: src/openfl/display/SimpleButton.hx (hunk 3 above)
Comment (NenadBojkovski) [reply src/openfl/display/SimpleButton.hx]: When you look at the code it at first sight, it looks like it can never be null. There also methods that are using it  without null checks like `getBounds`.  But current state can be null. Will update the code

File: src/openfl/display/SimpleButton.hx [hunk 4]
```diff
@@ -101,6 +101,19 @@ class SimpleButton extends InteractiveObject {
...
 	
+	private override function __forceRenderDirty ():Void {
+		
+		super.__forceRenderDirty();
+		
+		if (__currentState != null) {
+			
+			__currentState.__forceRenderDirty ();
+			
+		}
```
Comment (Neu2o) [inline src/openfl/display/SimpleButton.hx:112]: As discussed, the `hitTestState` doesn't need to be set to dirty. It is not rendered and thus not effected by something like `devicePixelRatio`. Nice `;)`

**Review by Neu2o — APPROVED** [review]
